wizard/class-instant-articles-wizard-state.php

```diff
@@ -0,0 +1,352 @@
+			return self::$transition_vectors[ $from ][ $to ];
+		}
+		return null;
+	}
+
+
+	/**
+	 * Executes the transition from the current state to a new state
+	 *
+	 * @param string $new_state Constant for the final state of the transition
+	 * @param mixed[] $params Parameters for the transition
+	 * @throws Instant_Articles_Invalid_Wizard_Transition_Exception if an invalid transition is attempted
+	 */
+	public static function do_transition( $new_state, $params = array() ) {
+		$current_state = self::get_current_state();
+		$transition = self::get_transition( $current_state, $new_state );
+
+		if ( $transition === null ) {
+			throw new Instant_Articles_Invalid_Wizard_Transition_Exception( $current_state, $new_state );
+		}
+
+		switch ( $transition ) {
+			case self::TRANSITION_START_WIZARD:
+				return self::transition_start_wizard();
+
+			case self::TRANSITION_SET_UP_APP:
+				return self::transition_set_up_app( $params[ 'app_id' ], $params[ 'app_secret'], $params[ 'user_access_token' ] );
+
+			case self::TRANSITION_SELECT_PAGE:
+				// TODO: validate params
+				return self::transition_select_page( $params[ 'page_id'] );
+			case self::TRANSITION_SELECT_STYLE:
+				// TODO: validate params
+				return self::transition_select_style();
+			case self::TRANSITION_EDIT_APP:
+				return self::transition_edit_app();
+			case self::TRANSITION_EDIT_PAGE:
+				return self::transition_edit_page();
+			case self::TRANSITION_EDIT_STYLE:
+				return self::transition_edit_style();
+		}
+	}
+
+	/**
+	 * Claims the URL for this site.
+	 */
+	public static function claim_url() {
+		$fb_app_settings = Instant_Articles_Option_FB_App::get_option_decoded();
+		$fb_page_settings = Instant_Articles_Option_FB_Page::get_option_decoded();
+
+		if ( ! $fb_app_settings[ 'app_id' ] ) {
+			throw new InvalidArgumentException( 'Missing app_id for claiming the URL.' );
+		}
+		if ( ! $fb_app_settings[ 'app_secret' ] ) {
+			throw new InvalidArgumentException( 'Missing app_secret for claiming the URL.' );
+		}
+		if ( ! $fb_page_settings[ 'page_access_token' ] ) {
+			throw new InvalidArgumentException( 'Missing page_access_token for claiming the URL.' );
+		}
+		if ( ! $fb_page_settings[ 'page_id' ] ) {
+			throw new InvalidArgumentException( 'Missing page_id for claiming the URL.' );
+		}
+
+		$client = Client::create(
+			$fb_app_settings[ 'app_id' ],
+			$fb_app_settings[ 'app_secret' ],
+			$fb_page_settings[ 'page_access_token' ],
+			$fb_page_settings[ 'page_id' ]
+		);
+
+
+		// We need the home URL without the protocol for claiming
+		$url = preg_replace( '/^https?:\/\//i', '', esc_url_raw( home_url( '/' ) ) );
+
+		try {
+			$client->claimURL( $url );
+		} catch (Exception $e) {
+			// Here we override the error message to give an actionable
+			// instruction to the user that is specific for WordPress.
+			throw new Exception("Could not claim the URL for the selected page, please make sure your site is publicly available.");
```

Comment:
We do capture this exception on the call site, so it never leads to a fatal error.